src/ucs/arch/x86_64/cpu.c

```diff
@@ -292,4 +442,102 @@ void ucs_cpu_init()
 #endif
 }
 
-#endif
+ucs_status_t ucs_arch_get_cache_size(ucs_cpu_cache_type_t type, size_t *value)
+{
+    x86_cache_line_reg_info_t cache_info;
+    x86_cache_line_reg_info_t line_info;
+    x86_cpu_registers reg;
+    x86_cpu_version_t version;
+    uint32_t sets;
+    uint32_t i, t, r, l4;
+    uint32_t max_iter = 1;
+    uint32_t model, family;
+
+    /* Get CPU model/family */
+    ucs_x86_cpuid(X86_CPUID_GET_MODEL, &version.reg, &reg.ebx, &reg.ecx, &reg.edx);
+
+    ucs_x86_cpuid(X86_CPUID_GET_BASE_VALUE, &reg.eax, &reg.ebx, &reg.ecx, &reg.edx);
+    if ((memcmp(reg.id, X86_CPUID_GENUEINTEL, sizeof(X86_CPUID_GENUEINTEL) - 1)) ||
+        (reg.eax < X86_CPUID_GET_CACHE_INFO)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    for(i = 0; i < max_iter; i++) {
+        ucs_x86_cpuid(X86_CPUID_GET_CACHE_INFO, &reg.eax, &reg.ebx, &reg.ecx, &reg.edx);
+
+        if (i == 0) { /* on first iteration get max iteration number */
+            max_iter     = reg.max_iter;
+            reg.max_iter = 0; /* mask iteration register from processing */
+        }
+
+        for (r = 0; r < sizeof(reg.reg) / sizeof(reg.reg[0]); r++) {
+            if (ucs_test_all_flags(reg.reg[r].value, X86_CPU_CACHE_RESERVED)) {
+                continue;
+            }
+
+            for (t = 0; (t < sizeof(reg.reg[r].tag) / sizeof(reg.reg[r].tag[0])) && (reg.reg[r].tag[t] != 0); t++) {
+                switch(reg.reg[r].tag[t]) {
+                case X86_CPU_CACHE_TAG_L1_ONLY:
+                    if ((type == UCS_CPU_CACHE_L2) || (type == UCS_CPU_CACHE_L3)) {
+                        return UCS_ERR_UNSUPPORTED;
+                    }
+                    break;
+                case X86_CPU_CACHE_TAG_LEAF4:
+                    for (l4 = 0;; l4++) {
+                        asm volatile("cpuid"
+                                    : "=a"(cache_info), "=b"(line_info), "=c"(sets), "=d"(reg.edx)
+                                    : "0"(X86_CPUID_GET_LEAF4_INFO), "2"(l4));
+
+                        if (cache_info.type == 0) {
+                            /* we are done - nothing found */
+                            break;
+                        }
+
+                        if ((cache_info.level == x86_cpu_cache[type].level) &&
+                            (cache_info.type  == x86_cpu_cache[type].type)) {
+                            /* found it */
+                            *value = (line_info.associativity + 1) *
+                                     (line_info.partitions + 1)    *
+                                     (line_info.size + 1)          *
+                                     (sets + 1);
+                            return UCS_OK;
+                        }
+                    }
+                    break;
+                default:
+                    if (reg.reg[r].tag[t] == 0x49) {
```

Comment:
I believe you an omit the special case of Xeon MP, which was the Pentium IV-era CPU, clearly irrelevant for UCX applications.

Reply:
not sure got your idea... do you mean remove this special-processor branch?

Reply:
> not sure got your idea... do you mean remove this special-processor branch?

Yes, ignore the special case for Xeon MP and always interpret 0x49 as '2nd-level cache: 4 MByte, 16-way set associative, 64 byte line size'.

Reply:
ok, removed